### PR TITLE
postgres: add ability to terminate SSL at Envoy

### LINKED ISSUE #10942: Postgres filter: implement Postgres SSL termination and monitoring
## Previous work

This issue elaborates on the general design of a Postgres filter proposed in #9107.

## Background

Encrypting the communications with the database is a hard requirement in many environments. And while cryptography is currently very fast on modern hardware, it still imposes some penalty where it is executed. Particularly, establishing SSL connections is expensive.

RDBMs like Postgres have a primary-replicas architecture, where the former is the only instance that takes writes. Offloading SSL from the primary instance can help reduce the workload, and increase the room for vertical scalability of these services. Similarly, some connection poolers frequently used in combination with Postgres, like PgBouncer, are single-threaded and may saturate the CPU soon if they are dealing with frequent SSL connection establishment.

Those are good enough reasons to implement **SSL termination in the Postgres filter for the Envoy proxy**. It is very convenient also because certificate management can also be offloaded, and potentially handled via [Envoy APIs](https://www.envoyproxy.io/docs/envoy/latest/api-v3/admin/v3/certs.proto) (and existing tools that leverage them), without having to change Postgres configuration.

Moreover, [the current version of the Postgres Envoy Filter implements several traffic inspection metrics](https://www.envoyproxy.io/docs/envoy/latest/configuration/listeners/network_filters/postgres_proxy_filter#config-network-filters-postgres-proxy) that are useful for monitoring. But they (obviously) cannot peek into the SSL traffic, exposing these metrics only for unencrypted connections (as of today).

It might be argued that the CPU offloading advantage is not present in scenarios where Envoy is deployed as a sidecar within the same Pod as Postgres. While true, it doesn’t neglect anyway the advantages of monitoring metrics of the encrypted traffic, separation of concerns and API-based management of certificates.

## Goals

- Implement Postgres SSL Termination at the existing Envoy Proxy’s Postgres filter.
- Adapt the filter to expose the same metrics it exposes now for unencrypted traffic also for the encrypted traffic terminated at Envoy.
- Provide SSL configuration capabilities both via the config file and the Envoy APIs.
- Optional (or next phase refinements). Provide (equivalent) support to Postgres’ advanced SSL configuration capabilities, like [ssl_ciphers](https://postgresqlco.nf/en/doc/param/ssl_ciphers/), [ssl_ecdh_curve](https://postgresqlco.nf/en/doc/param/ssl_ecdh_curve/) or [ssl_min_protocol_version](https://postgresqlco.nf/en/doc/param/ssl_min_protocol_version/), among several others.

## Non-goals

- Support encrypted communications from Envoy to the upstream Postgres server. It is assumed that this communication will happen unencrypted.
- Support client authentication via SSL certificates (this would be a goal for a future issue).

## Implementation notes

Envoy does support SSL termination at the TCP layer. However Postgres SSL support does not happen at the TCP level, but rather at the application layer (Frontend/Backend Protocol). This questions the amount of existing infrastructure that may be reused. In any case, the same SSL library that Envoy uses, BoringSSL, will be used. It is a key requirement that TLS levels and general compatibility with Postgres SSL are appropriately tested.

The following diagram (note that the Postgres terminology of the FrontEnd-BackEnd protocol is used, instead of the usual Downstream/Upstream at Envoy) illustrates how the encrypted/unencrypted flows may work:

Diagram-PGClient_Envoy_PGServer-v1

Note that the “fake response” happens after SSL Request, where we will “imitate” the backend sending back the byte “S” to the frontend.

Ideally, we should leverage as much as possible all the SSL infrastructure available already in Envoy, and not create different configuration files/keys/APIs. Here is an example of how Envoy TLS is currently configured:

```yaml
tls_context:
  common_tls_context:
    tls_certificates:
      - certificate_chain:
          filename: "/etc/example-com.crt"
        private_key:
          filename: "/etc/example-com.key"
```

## Limitations
- Potentially, some SSL versions or encryption mechanisms may not work, and a reduced set of options may be exposed. This should be fine as Postgres clients negotiate with the server the mechanisms to use. There could be, potentially, some limitations with some specific clients, but it is not expected to be a relevant issue.

- [SCRAM authentication with channel binding](https://www.postgresql.org/docs/current/sasl-authentication.html#SASL-SCRAM-SHA-256) may not be used when proxying through Envoy, as the Postgres server will not be running in SSL mode, and Postgres implementation of channel binding uses tls-server-end-point.

## References
- [https://openssl.org](https://openssl.org)
- [https://aticleworld.com/ssl-server-client-using-openssl-in-c](https://aticleworld.com/ssl-server-client-using-openssl-in-c)
- [https://github.com/postgres/postgres/blob/master/src/interfaces/libpq/fe-secure-openssl.c](https://github.com/postgres/postgres/blob/master/src/interfaces/libpq/fe-secure-openssl.c)
- [https://github.com/postgres/postgres/blob/master/src/backend/libpq/be-secure-openssl.c](https://github.com/postgres/postgres/blob/master/src/backend/libpq/be-secure-openssl.c)
- [https://github.com/envoyproxy/envoy/tree/master/source/extensions/transport\_sockets/tls](https://github.com/envoyproxy/envoy/tree/master/source/extensions/transport_sockets/tls)
- [https://www.postgresql.org/docs/current/protocol-flow.html\#id-1.10.5.7.11](https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.11)
- [https://www.postgresql.org/docs/current/ssl-tcp.html](https://www.postgresql.org/docs/current/ssl-tcp.html)
- [https://www.postgresql.org/docs/current/libpq-ssl.html](https://www.postgresql.org/docs/current/libpq-ssl.html)

### PR DESCRIPTION
**Pull request by cpakulski**
Commit Message:
Adds ability to use _starttls_ transport socket to terminate SSL at Envoy and pass unencrypted traffic upstream to Postgres server.

Additional Description:
Risk Level: Low
Testing: Added unit and integration tests.
Docs Changes: Yes.
Release Notes: Yes. 
Fixes #10942

### CODE REVIEW

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:41]: Can you elaborate on why we need to do this via StartTls if this is static configuration? Why can't the listener just be configured with TLS support?

Comment (htuch) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:41]: I see, this is covered in https://github.com/envoyproxy/envoy/issues/10942.

Comment (cpakulski) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:41]: I was just writing the reply. Yes, everything is captured in #10942. TLS is negotiated in-band and protocol starts in clear-text.

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:40]: Just make this a regular `bool` and no need to state the default as it's implied.

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (htuch) [inline api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:36]: If you have a link to the Postgres StartTls protocol that could be useful here.

Comment (fabriziomello) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:36]: This one: https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.11

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (cpakulski) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto]: Modified to use `bool`. Thanks. 
Quick question: previously when I run `proto_format.sh fix` it propagated changes forward to v4alpha. But not this time. Is it because api version here is v3alpha? Maybe I should change it to v3?

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (cpakulski) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:36]: Added link, but used _rel13_ instead of _latest_. As _latest_ moves forward the link may become invalid.

Comment (fabriziomello) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:36]: It wont... the use of `current` keyword on docs URL is a common pattern in PostgreSQL community to lead you to the latest stable release. We do major releases once per year and what we do here with protocol decoding, etc don't change for last 17 years, so IMHO is better to use `current` instead of point to some specific version.

**Comment by fabriziomello** [comment]
@cpakulski As you already know I did a lot of tests against this new implementation using Postgres regression test suite. But need to test more cases: replication and authentication using client certificates (testing it today).

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (cpakulski) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto:36]: OK. I will change to use `current` when I get more comments on PR.

**Review by htuch — COMMENTED** [review]
/lgtm api

File: api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto
Comment (htuch) [reply api/envoy/extensions/filters/network/postgres_proxy/v3alpha/postgres_proxy.proto]: The reason it's not doing this is that there is no type dependency that would cause a breaking change in v4 and we don't automatically bump such protos. We should, but that's a tooling issue you don't need to deal with, I'll comment in https://github.com/envoyproxy/envoy/issues/12841.

File: source/extensions/filters/network/postgres_proxy/postgres_filter.cc
Comment (dio) [inline source/extensions/filters/network/postgres_proxy/postgres_filter.cc:204]: Sorry, this is mostly my ignorance on the protocol, here: https://www.postgresql.org/docs/current/protocol-flow.html#id-1.10.5.7.11 we expect 'S' and 'N'? Where in the code that we handle 'N' response?

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (dio) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:211]: Add this ref: https://www.postgresql.org/docs/current/protocol-message-formats.html? SSLRequest "magic number" (code). The value is chosen to contain 1234 in the most significant 16 bits, and 5679 in the least significant 16 bits.

0x04D2 = 1234
0x162F = 5679

File: source/extensions/filters/network/postgres_proxy/postgres_filter.cc
Comment (fabriziomello) [reply source/extensions/filters/network/postgres_proxy/postgres_filter.cc:204]: Because here we need to mimic Postgres server sending back to client that we support SSL so the client is able to start the handshake.

And if not exists a transport socket then the connection will be finished.

Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_filter.cc:204]: @dio Thanks for reviewing. The logic here is that the postgres filter "captures" the initial request from the client and does not allow it to reach the server. Then it sends a response 'S', which means "let's switch to SSL". After the response has been transmitted, it calls starttls socket to switch to TLS/SSL. The call to switch to TLS may fail (configuration, wrong certs, etc). Theoretically when switch to TLS fails, the postgres filter should transmit 'N', which means "no, I do not allow to switch to TLS". But the filter has already transmitted 'S' and if switch to TLS fails, postgres filter will close the TCP connection to the client.

In other words, we do not send 'N', but close connection when things go wrong.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:211]: I will add this. Thanks.

File: source/extensions/filters/network/postgres_proxy/postgres_filter.cc
Comment (fabriziomello) [reply source/extensions/filters/network/postgres_proxy/postgres_filter.cc:204]: @dio this [image](https://user-images.githubusercontent.com/1239643/80261846-e5129d80-868b-11ea-995d-0a55ae96ce87.png) from original issue describes this behavior. For sure it don't cover the STARTTLS stuff because at that time we have no idea about all Envoy internals we need to change/improve to implement this feature.

Comment (dio) [reply source/extensions/filters/network/postgres_proxy/postgres_filter.cc:204]: Thank you for the explanation!

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:212]: Updated.

**Comment by fabriziomello** [comment]
@cpakulski @dio @htuch here is the first WIP version of repository to battle test Envoy Postgres Filter using [PostgreSQL regression tests](https://www.postgresql.org/docs/current/regress.html):

https://github.com/fabriziomello/envoy-postgres-regression/runs/1769320312?check_suite_focus=true

The execution above is over the current branch.

**Comment by cpakulski** [comment]
@htuch I need another LGTM for api. I only updated url pointing to documentation in proto file since https://github.com/envoyproxy/envoy/pull/14634#pullrequestreview-567688655. 
API itself was not changed.
Thank you.

**Comment by htuch** [comment]
API LGTM, will defer to @dio for implementation.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (dio) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:219]: Should we log the returned value as well?

File: source/extensions/filters/network/postgres_proxy/postgres_filter.cc
Comment (dio) [inline source/extensions/filters/network/postgres_proxy/postgres_filter.cc:212]: Nit. s/tls/TLS for consistency

**Review by dio — COMMENTED** [review]
Thank you for working on this! I learned a lot by reading the code. Sorry for the delay. I Will do another pass, but meanwhile, flushing some comments and questions.

File: source/extensions/filters/network/postgres_proxy/config.cc
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/config.cc:24]: Done.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.h
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.h]: Added comments to each enum.

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (cpakulski) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:218]: I removed the log from here and added it to `onSSLRequest` method. When terminateSSL option is enabled, log will indicate if it was successful or not. If terminateSSL is disabled, nothing will be logged and messages will pass in encrypted form without decoding.

File: docs/root/version_history/current.rst
Comment (dio) [inline docs/root/version_history/current.rst:64]: Bad merging?

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.cc
Comment (dio) [reply source/extensions/filters/network/postgres_proxy/postgres_decoder.cc:218]: Cool. Thanks!

File: source/extensions/filters/network/postgres_proxy/postgres_decoder.h
Comment (dio) [inline source/extensions/filters/network/postgres_proxy/postgres_decoder.h:63]: Thank you @cpakulski!

File: test/extensions/filters/network/postgres_proxy/postgres_integration_test.cc
Comment (dio) [inline test/extensions/filters/network/postgres_proxy/postgres_integration_test.cc:146]: Super nit, newline.

**Review by dio — COMMENTED** [review]
Looks good. Observed bad merging in the version history.

**Comment by fabriziomello** [comment]
And all Postgres regression are OK with last commits of this PR: https://github.com/fabriziomello/envoy-postgres-regression/runs/1810089899?check_suite_focus=true

File: docs/root/version_history/current.rst
Comment (cpakulski) [reply docs/root/version_history/current.rst]: Corrected. Thanks for catching it!

File: test/extensions/filters/network/postgres_proxy/postgres_integration_test.cc
Comment (cpakulski) [reply test/extensions/filters/network/postgres_proxy/postgres_integration_test.cc:147]: Added.

**Review by dio — APPROVED** [review]
Thanks, awesome work! @fabriziomello I think we need to pull in the regression test to be part of the envoy test suite, as mentioned before, probably verify_examples is one of the possible places to do so.

**Comment by cpakulski** [comment]
Thanks @dio for reviewing and valuable comments!

**Comment by fabriziomello** [comment]
> Thanks, awesome work! @fabriziomello I think we need to pull in the regression test to be part of the envoy test suite, as mentioned before, probably verify_examples is one of the possible places to do so.

Are you meant here? https://github.com/envoyproxy/envoy/tree/main/examples/postgres

**Comment by dio** [comment]
@fabriziomello yes.

**Review by lizan — APPROVED** [review]

**Pull request merged**
